lib/sotest.nix

```diff
@@ -94,4 +97,60 @@ rec {
       exec = builtins.concatStringsSep " " ([ kernel ] ++ kernelParams);
       load = [ initrd ];
     };
+
+  # Given a test run attribute set, this function is used to build a bundle consisting of a project
+  # config in YAML format, and a ZIP archive containing exactly the referenced test files. This
+  # function will only work for test run descriptions that exclusively use content of the nix store
+  # for their `exec`, `load`, and `extra_files` attributes.
+  #
+  # Only paths at the beginning of an entry will be included in the resulting zip archive. In
+  # particular, a boot item with an entry like
+  #
+  # exec = "/nix/store/<hash>-linux/bzImage init=/nix/store/<hash>-linux-system/init";
+  #
+  # will NOT cause /nix/store/<hash>-linux-system/init to be included in the zip archive. The
+  # corresponding `exec` entry in the resulting YAML file will contain:
+  #
+  # exec: "<hash>-linux/bzImage init=/nix/store/<hash>-linux-system/init"
+  #
+  # Limitations:
+  #
+  # - The parsing logic for entries is limited to `pkgs.lib.splitString " "`, which is why this
+  #   function will not work correctly for paths that do not contain whitespace.
+  mkProjectBundle = testRunConfig:
+    let
+      # nixpkgs has a function `updateManyAttrsByPath`, which almost does what we need, but doesn't
+      # support optional attributes in the way we need it, which is why we write our own little
+      # helper here.
+      updateAttr = key: default: update: attrs: attrs // {
+        "${key}" = if builtins.hasAttr key attrs then update (builtins.getAttr key attrs) else default;
+      };
+      error = attrname: name:
+        throw "Missing required attribute \"${attrname}\" in boot item ${name}";
+      stripPrefixFromBootItem = boot_item:
+        updateAttr "exec" (error "exec" boot_item.name) (pkgs.lib.removePrefix "/nix/store/")
+          (updateAttr "load" [ ] (map (pkgs.lib.removePrefix "/nix/store/"))
+            (updateAttr "extra_files" [ ] (map (pkgs.lib.removePrefix "/nix/store/"))
+              boot_item
+            ));
+      collectEntries = boot_item: [ boot_item.exec ] ++ (boot_item.load or [ ]) ++ (boot_item.extra_files or [ ]);
+      strippedConfig = updateAttr "boot_items" [ ] (map stripPrefixFromBootItem) testRunConfig;
+      paths =
+        # splitString will always return a list with at least one element, which is why
+        # builtins.head is safe to use here.
+        map builtins.head
+          (map (pkgs.lib.strings.splitString " ")
+            (builtins.concatMap collectEntries strippedConfig.boot_items));
+      configYaml = asYAMLFile (builtins.toJSON strippedConfig);
+    in
+    pkgs.runCommandNoCC "sotest-bundle"
+      {
+        preferLocalBuild = true;
```

Comment:
it seems like local build is not a prefered thing because if i really want this bundle, then it would be much faster to download the final minimized zip ball instead of having to download (or even build) all the packages it depends on, right?

preferLocalBuild is usually used where it's easier to build the derivation itself because it is executed faster than a cache query and download.

Reply:
As far as I understand (and from what I observed), `preferLocalBuild = true;` will not prevent inputs of this derivation to be downloaded on my machine. That happens in any case it cannot directly substitute the result of the derivation. But what it prevents is (after downloading / building all the inputs), having a remote builder put them into an archive, and then downloading that archive again. This is especially painful when the inputs and the resulting archive is considerably large.

Substitution from already built archives in a binary cache should still work, because this is controlled by `allowSubstitutes`.

Reply:
ok, my fault. i accidentally swapped the meanings of those two flags in my head so my initial comment just makes no sense in this regard. :-D